### PR TITLE
[rush] Add warning when using globalIgnoredOptionalDependencies in < pnpm 9.0.0

### PR DESCRIPTION
**Pull request by witcher112**
## Summary

"globalIgnoredOptionalDependencies" doesn't work if Rush monorepo is not configured with pnpm 9+.

## Details

I've added mentions about it in the schema and a warning during rush install/update.

image

## Impacted documentation

https://rushjs.io/pages/configs/pnpm-config_json/

### CODE REVIEW

File: libraries/rush-lib/src/logic/installManager/InstallHelpers.ts [hunk 1]
```diff
@@ -71,6 +72,18 @@ export class InstallHelpers {
...
 
       if (pnpmOptions.globalIgnoredOptionalDependencies) {
+        if (rushConfiguration.rushConfigurationJson.pnpmVersion !== undefined && semver.lt(rushConfiguration.rushConfigurationJson.pnpmVersion, '9.0.0')) {
+          // eslint-disable-next-line no-console
+          console.warn(
+            Colorize.yellow(
+              `Your version of pnpm ${rushConfiguration.rushConfigurationJson.pnpmVersion} ` +
+                `doesn't support "globalIgnoredOptionalDependencies". Please consider upgrading the ` +
+                `"pnpmVersion" setting in ${RushConstants.rushJsonFilename} to 9.0.0 or higher ` +
+                Colorize.bold('(please note that pnpm 9+ is not officially supported yet at the moment)')
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/InstallHelpers.ts:82]:
```suggestion
              `Your version of pnpm (${rushConfiguration.rushConfigurationJson.pnpmVersion}) ` +
                `doesn't support the "globalIgnoredOptionalDependencies" field in ${RushConstants.<whatever this is>}/${RushConstants.<pnpm-config.json constant>}. ` +
                `Remove this field or upgrade to pnpm 9.`
```

I don't think it's necessary to mention that pnpm 9 isn't officially supported yet in this particular error message.

File: libraries/rush-lib/src/logic/installManager/InstallHelpers.ts [hunk 2]
```diff
@@ -71,6 +72,18 @@ export class InstallHelpers {
       }
 
       if (pnpmOptions.globalIgnoredOptionalDependencies) {
+        if (rushConfiguration.rushConfigurationJson.pnpmVersion !== undefined && semver.lt(rushConfiguration.rushConfigurationJson.pnpmVersion, '9.0.0')) {
+          // eslint-disable-next-line no-console
+          console.warn(
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/InstallHelpers.ts:77]: Use `terminal`.

File: common/changes/@microsoft/rush/warn-about-ignored-optional-dependencies-pnpm-pre-9_2024-11-09-12-28.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Add warning when using globalIgnoredOptionalDependencies in < pnpm 9.0.0",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/warn-about-ignored-optional-dependencies-pnpm-pre-9_2024-11-09-12-28.json:5]:
```suggestion
      "comment": "Add warning when the `globalIgnoredOptionalDependencies` property is specified in `common/config/rush/pnpm-config.json` and the repo is configured to use pnpm <9.0.0.",
```

**Review by iclanton — COMMENTED** [review]
This should be an error, not a warning. If someone configures something and that configuration isn't going to do anything, we usually error in that case.

File: libraries/rush-lib/src/logic/installManager/InstallHelpers.ts (hunk 2 above)
Comment (witcher112) [reply libraries/rush-lib/src/logic/installManager/InstallHelpers.ts:77]: The thing is that there's no `terminal` accessible in this scope. Should I rework the code a bit so it's passed down and accessible here or create a new instance of it?

Comment (iclanton) [reply libraries/rush-lib/src/logic/installManager/InstallHelpers.ts:77]: If that's easy enough to do, go for it, otherwise don't worry about it.

**Review by iclanton — APPROVED** [review]

**Pull request merged**